Libraries/MLXLMCommon/LanguageModel.swift

```diff
@@ -0,0 +1,220 @@
+// Copyright © 2024 Apple Inc.
+
+import Foundation
+import Hub
+import MLX
+import MLXNN
+import Tokenizers
+
+/// Time/Height/Width struct to represent information about input images.
+public struct THW: Sendable {
+
+    public let t: Int
+    public let h: Int
+    public let w: Int
+
+    public init(_ t: Int, _ h: Int, _ w: Int) {
+        self.t = t
+        self.h = h
+        self.w = w
+    }
+
+    public var values: (Int, Int, Int) {
+        (t, h, w)
+    }
+
+    public var product: Int { t * h * w }
+}
+
+/// Representation of ``LanguageModel`` input.
+///
+/// This can contain text (tokens), prepared images (`MLXArray`), or other media as
+/// needed.  ``LMInput`` is produced by ``UserInputProcessor`` in response
+/// to ``UserInput``.
+///
+/// The ``ModelContext`` holds the ``UserInputProcessor`` associated with a
+/// ``LanguageModel``.
+public struct LMInput {
```

Comment:
LMInput and LMOutput are perhaps the two largest changes: this encapsulates the different values we need to pass as input and output and allows us to add new values (e.g. video frames) over time without changing the language model API.

Previously these were bare MLXArrays.